[Instrumentation] Mark instrumented calls as implicit

serge-sans-paille opened the pull request:
Make sure that instrumentation code doesn't interferes with with prologue endpoint computation when generating debug information.

Fix #54873

jmorse reviewed (commented):
Hola; this looks like a good fix, however I wonder whether existing facilities for implicit code would be sufficient to achieve this. Nowadays, by convention, a DILocation with a line-number of zero is interpreted as being a compiler-generated source location and is ignored by various pieces of LLVM (and DWARF consumers too). Would using that achieve your aims instead?

Specifically, I gave this a quick spin:

```
diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index 6a89cee9aaf6..48d00cfb9671 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -949,7 +949,7 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
   DebugLoc EntryLoc;
   if (IsEntryBB) {
     if (auto SP = F.getSubprogram())
-      EntryLoc = DILocation::get(SP->getContext(), SP->getScopeLine(), 0, SP);
+      EntryLoc = DILocation::get(SP->getContext(), 0, 0, SP);
     // Keep static allocas and llvm.localescape calls in the entry block.  Even
     // if we aren't splitting the block, it's nice for allocas to be before
     // calls.
```

and the reproducer from the linked ticket/issue shifted the prologue_end flag back to the desired position. As a side-effect the instrumentation code then can't be stepped on (a feature, in my view).

Disclosure: I'd prefer to avoid more users of isImplicitCode, as it'd be quite nice to repurpose it in the future, and it's little-used right now.

serge-sans-paille committed:
[Instrumentation] Do not attache line number to instrumented call debuginfo

Doing so makes sure instrumentation code doesn't interferes with with
prologue endpoint computation when generating debug information, as
prologue_endpoint computation skips over such instructions.

Fix #54873

serge-sans-paille commented:
The change dates back to 201733b7f0b4fee6120dc1e0e43df90a860a34be, and I agree with the simplicity of your approach, let me update the diff.

serge-sans-paille commented:
@jmorse : updated!

jmorse commented on llvm/test/Instrumentation/SanitizerCoverage/coverage-dbg.ll:
```diff
@@ -16,7 +16,7 @@
 
 ; Test that __sanitizer_cov_trace_pc_guard call has !dbg pointing to the opening { of A::f().
 ; CHECK: call void @__sanitizer_cov_trace_pc_guard(ptr{{.*}}) #{{.*}}, !dbg [[A:!.*]]
```
This comment will want updating as it's now not true I guess; it's unclear to me whether this indicates there are expectations that the coverage call gets a "real" source location?

jmorse commented on llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp:
```diff
@@ -949,7 +949,7 @@ void ModuleSanitizerCoverage::InjectCoverageAtBlock(Function &F, BasicBlock &BB,
   if (IsEntryBB) {
     if (auto SP = F.getSubprogram())
-      EntryLoc = DILocation::get(SP->getContext(), SP->getScopeLine(), 0, SP);
+      EntryLoc = DILocation::get(SP->getContext(), 0, 0, SP);
```
IMHO this warrants a comment along the lines of "Use a compile-generated source-location for all coverage calls", to aid future readers

jmorse reviewed (approved):
This LGTM with a couple of nits;  it's not clear to me whether using line-zero on coverage-instrumentation calls breaks some convention or developer expectation, as coverage-instrumentation isn't an area I'm familiar with.

serge-sans-paille committed:
fixup! [Instrumentation] Do not attache line number to instrumented call debuginfo

khuey commented:
This needs to be tested against actual users of this API before it's merged. The example given in the docs (https://clang.llvm.org/docs/SanitizerCoverage.html)

```c
extern "C" void __sanitizer_cov_trace_pc_guard(uint32_t *guard) {
  if (!*guard) return;  // Duplicate the guard check.
  // If you set *guard to 0 this code will not be called again for this edge.
  // Now you can get the PC and do whatever you want:
  //   store it somewhere or symbolize it and print right away.
  // The values of `*guard` are as you set them in
  // __sanitizer_cov_trace_pc_guard_init and so you can make them consecutive
  // and use them to dereference an array or a bit vector.
  void *PC = __builtin_return_address(0);
  char PcDescr[1024];
  // This function is a part of the sanitizer run-time.
  // To use it, link with AddressSanitizer or other sanitizer.
  __sanitizer_symbolize_pc(PC, "%p %F %L", PcDescr, sizeof(PcDescr));
  printf("guard: %p %x PC %s\n", guard, *guard, PcDescr);
}
```

Seems very likely to break (__sanitizer_symbolize_pc will end up finding the line 0 location with the instrumentation sequence in #54873).

vitalybuka commented:
> This needs to be tested against actual users of this API before it's merged. The example given in the docs (https://clang.llvm.org/docs/SanitizerCoverage.html)
> 
> ```c
> extern "C" void __sanitizer_cov_trace_pc_guard(uint32_t *guard) {
>   if (!*guard) return;  // Duplicate the guard check.
>   // If you set *guard to 0 this code will not be called again for this edge.
>   // Now you can get the PC and do whatever you want:
>   //   store it somewhere or symbolize it and print right away.
>   // The values of `*guard` are as you set them in
>   // __sanitizer_cov_trace_pc_guard_init and so you can make them consecutive
>   // and use them to dereference an array or a bit vector.
>   void *PC = __builtin_return_address(0);
>   char PcDescr[1024];
>   // This function is a part of the sanitizer run-time.
>   // To use it, link with AddressSanitizer or other sanitizer.
>   __sanitizer_symbolize_pc(PC, "%p %F %L", PcDescr, sizeof(PcDescr));
>   printf("guard: %p %x PC %s\n", guard, *guard, PcDescr);
> }
> ```
> 
> Seems very likely to break (__sanitizer_symbolize_pc will end up finding the line 0 location with the instrumentation sequence in #54873).


I wondering if https://clang.llvm.org/docs/SanitizerCoverage.html#id16 will still work?

vitalybuka reviewed (changes requested):
Removing from my pending review, as there are open concerns.
